keps/sig-cloud-provider/20190722-ccm-instance-metadata.md

```diff
@@ -0,0 +1,242 @@
+
+## Motivation
+
+Before CCM, kubelet supports getting Node information by the cloud provider's instance metadata service. This includes:
+
+- NodeName
+- ProviderID
+- NodeAddresses
+- InstanceType
+- AvailabilityZone
+
+Instance metadata service could help to reduce API throttling issues and the node's initialization duration. This is especially helpful for large clusters. But with CCM, this is not possible anymore because the above functionality has been moved to CCM.
+
+Take Azure cloud provider for example:
+
+- According to Azure documentation [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits), for each Azure subscription and tenant, Resource Manager allows up to **12,000 read requests per hour** and 1,200 write requests per hour. That means, on average only 200 read requests could be sent per minute.
+- For different Azure APIs, there’re also additional rate limits based on different durations. For example, there are 3Min and 30Min read limits for VMSS APIs (the numbers below are only for reference since they’re not officially [documented](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits)):
+  - Microsoft.Compute/HighCostGetVMScaleSet3Min;200
+  - Microsoft.Compute/HighCostGetVMScaleSet30Min;1000
+  - Microsoft.Compute/VMScaleSetVMViews3Min;5000
+
+Based on those rate limits, getting node’s information for a 5000 cluster may need hours.  Things would be much worse for multiple clusters in the same tenant and subscription.
+
+So the proposal aims to add IMDS support back with CCM, so that the kubernetes cluster could still be scaled to a large number of nodes.
+
+### Goals
+
+- Allow nodes to be initialized from IMDS.
+- Allow nodes to reconcile the node addresses from IMDS.
+
+### Non-Goals
+
+- Authentication and authorization for each provider implementations.
+- API throttling [issue](https://github.com/kubernetes/kubernetes/issues/60646) on route controller.
+
+## Proposal
+
+Same as kube-controller-manager, the [cloud provider interfaces](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L43) could be split into two parts:
+
+- Instance-level interfaces: Instances and Zones
+- Control-plane interfaces, e.g. LoadBalancer and Routes
+
+The control-plane interfaces are still kept in CCM (who’s name is cloud-controller-manager), and deployed on masters. For instance-level interfaces, a new daemonsets would be introduced and implement instance-level interfaces (who’s name is cloud-node-manager).
+
+With these changes, the whole node initialization workflow would be:
+
+- Kubelet specifying `--cloud-provider=external` will add a taint `node.cloudprovider.kubernetes.io/uninitialized` with an effect NoSchedule during initialization.
+- cloud-node-manager would initialize the node again with `Instances` and `Zones`.
```

Comment:
So `cloud-node-manager` only watches it's own node and after registration does nothing?

Reply:
See L76: reconcile the IP addresses would also be moved from ccm to cloud-node-manager.

I should probably also add the description here.